File: scripts/repair-weekly-summary-markup.py
import json
import re
from pathlib import Path

repo_root = Path(__file__).resolve().parents[1]


def strip_broken_html_tail(text: str) -> str:
    t = str(text or "")
    open_idx = t.rfind("<span")
    if open_idx != -1 and t.find("</span>", open_idx) == -1:
        t = t[:open_idx].strip()
    for pat in (
        r"<span[^>]*$",
        r"</?span[^>]*$",
        r"</?sp(?:an)?[^>]*$",
        r"<[^>]*$",
    ):
        t = re.sub(pat, "", t, flags=re.I)
    return t.strip()


def normalize_weekly_span_markup(text: str) -> str:
    t = re.sub(
        r"<span class='(num|highlight)'>([\s\S]*?)</span>",
        r'<span class="\1">\2</span>',
        str(text or ""),
        flags=re.I,
    )
    return strip_broken_html_tail(t)


def looks_truncated(text: str) -> bool:
    t = str(text or "")
    return bool(re.search(r"以<$|</sp(?:an)?$|<span[^>]*$|<[^>]*$", t, re.I))


def repair_structured(structured):
    truncated = False
    if not isinstance(structured, dict):
        return structured, truncated
    out = {}
    for key, items in structured.items():
        if not isinstance(items, list):
            continue
        out[key] = []
        for point in items:
            raw = str(point or "")
            if looks_truncated(raw):
                truncated = True
            out[key].append(normalize_weekly_span_markup(raw))
    return out, truncated


SECTION_KEYS = {
    "highlightsStructured": ["核心变化", "适用范围", "生效时间"],
    "impactsStructured": ["不利", "有利", "中性"],
    "actionsStructured": ["运营组", "客服组", "物流组"],
}


def flatten(structured, keys):
    flat = []
    if not isinstance(structured, dict):
        return flat
    for key in keys:
        for point in structured.get(key, []) or []:
            text = str(point or "").strip()
            if text:
                flat.append(f"{key}：{text[:200]}")
    return flat


def repair_file(rel_path: str):
    path = repo_root / rel_path
    rules = json.loads(path.read_text(encoding="utf-8-sig"))
    rules_truncated = 0
    for rule in rules:
        ai = rule.get("aiSummary")
        if not ai:
            continue
        rule_truncated = False
        for field, keys in SECTION_KEYS.items():
            if ai.get(field):
                ai[field], truncated = repair_structured(ai[field])
                rule_truncated |= truncated
        if ai.get("highlightsStructured"):
            ai["highlights"] = flatten(
                ai["highlightsStructured"], SECTION_KEYS["highlightsStructured"]
            )
        elif ai.get("highlights"):
            ai["highlights"] = [
                normalize_weekly_span_markup(x) for x in ai["highlights"]
            ]
        if ai.get("impactsStructured"):
            ai["impacts"] = flatten(
                ai["impactsStructured"], SECTION_KEYS["impactsStructured"]
            )
        elif ai.get("impacts"):
            ai["impacts"] = [normalize_weekly_span_markup(x) for x in ai["impacts"]]
        if ai.get("actionsStructured"):
            ai["actions"] = flatten(
                ai["actionsStructured"], SECTION_KEYS["actionsStructured"]
            )
        elif ai.get("actions"):
            ai["actions"] = [normalize_weekly_span_markup(x) for x in ai["actions"]]
        if ai.get("highlight"):
            ai["highlight"] = normalize_weekly_span_markup(ai["highlight"])
        if rule_truncated:
            rules_truncated += 1
            ai.pop("contentHash", None)
    path.write_text(
        "\ufeff" + json.dumps(rules, ensure_ascii=False, indent=2), encoding="utf-8"
    )
    return {
        "file": rel_path,
        "rules": len(rules),
        "need_resummarize": rules_truncated,
    }


if __name__ == "__main__":
    results = [
        repair_file("data/rules.json"),
        repair_file("public/data/rules.json"),
    ]
    print(json.dumps({"ok": True, "results": results}, ensure_ascii=False, indent=2))

File: scripts/repair-weekly-summary-markup.mjs
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { normalizeWeeklySpanMarkup } from "../src/services/llm/client.js";
import {
  flattenActionsStructured,
  flattenHighlightsStructured,
  flattenImpactsStructured
} from "../src/services/llm/prompts.js";

const repoRoot = path.join(path.dirname(fileURLToPath(import.meta.url)), "..");

function looksTruncated(text) {
  const t = String(text || "");
  return (
    /以<$/.test(t) ||
    /<\/sp(?:an)?$/i.test(t) ||
    /<span[^>]*$/i.test(t) ||
    /<[^>]*$/.test(t)
  );
}

function repairPoint(text) {
  return normalizeWeeklySpanMarkup(text);
}

function repairStructured(structured) {
  if (!structured || typeof structured !== "object") {
    return { structured, truncated: false };
  }
  let truncated = false;
  const next = {};
  for (const [key, items] of Object.entries(structured)) {
    if (!Array.isArray(items)) {
      continue;
    }
    next[key] = items.map((point) => {
      const raw = String(point || "");
      if (looksTruncated(raw)) {
        truncated = true;
      }
      return repairPoint(raw);
    });
  }
  return { structured: next, truncated };
}

function repairFlat(lines) {
  if (!Array.isArray(lines)) {
    return lines;
  }
  return lines.map((line) => repairPoint(line));
}

async function repairFile(relPath) {
  const filePath = path.join(repoRoot, relPath);
  const raw = await readFile(filePath, "utf8");
  const rules = JSON.parse(raw.replace(/^\uFEFF/, ""));
  let pointsFixed = 0;
  let rulesTruncated = 0;

  for (const rule of rules) {
    const ai = rule.aiSummary;
    if (!ai) {
      continue;
    }

    let ruleTruncated = false;

    if (ai.highlightsStructured) {
      const { structured, truncated } = repairStructured(ai.highlightsStructured);
      ai.highlightsStructured = structured;
      ruleTruncated ||= truncated;
    }
    if (ai.impactsStructured) {
      const { structured, truncated } = repairStructured(ai.impactsStructured);
      ai.impactsStructured = structured;
      ruleTruncated ||= truncated;
    }
    if (ai.actionsStructured) {
      const { structured, truncated } = repairStructured(ai.actionsStructured);
      ai.actionsStructured = structured;
      ruleTruncated ||= truncated;
    }

    if (ai.highlightsStructured) {
      ai.highlights = flattenHighlightsStructured(ai.highlightsStructured);
    } else if (ai.highlights) {
      ai.highlights = repairFlat(ai.highlights);
    }
    if (ai.impactsStructured) {
      ai.impacts = flattenImpactsStructured(ai.impactsStructured);
    } else if (ai.impacts) {
      ai.impacts = repairFlat(ai.impacts);
    }
    if (ai.actionsStructured) {
      ai.actions = flattenActionsStructured(ai.actionsStructured);
    } else if (ai.actions) {
      ai.actions = repairFlat(ai.actions);
    }
    if (ai.highlight) {
      ai.highlight = repairPoint(ai.highlight);
    }

    pointsFixed += 1;
    if (ruleTruncated) {
      rulesTruncated += 1;
      delete ai.contentHash;
    }
  }

  await writeFile(filePath, `\uFEFF${JSON.stringify(rules, null, 2)}`, "utf8");
  return { filePath, rules: rules.length, pointsFixed, rulesTruncated };
}

const targets = [
  "data/rules.json",
  "public/data/rules.json",
  "data/douyin/rules.json",
  "public/data/douyin/rules.json"
];
const results = [];
for (const rel of targets) {
  results.push(await repairFile(rel));
}

console.log(JSON.stringify({ ok: true, results }, null, 2));
